c/misra/src/rules/RULE-5-1/ExternalIdentifiersNotDistinct.ql

```diff
@@ -0,0 +1,39 @@
+/**
+ * @id c/misra/external-identifiers-not-distinct
+ * @name RULE-5-1: External identifiers shall be distinct
+ * @description Using nondistinct external identifiers results in undefined behaviour.
+ * @kind problem
+ * @precision very-high
+ * @problem.severity warning
+ * @tags external/misra/id/rule-5-1
+ *       correctness
+ *       maintainability
+ *       readability
+ *       external/misra/obligation/required
+ */
+
+import cpp
+import codingstandards.c.misra
+import codingstandards.cpp.Linkage
+
+class ExternalIdentifiers extends Declaration {
+  ExternalIdentifiers() { hasExternalLinkage(this) }
+}
+
+class ExternalIdentifiersLong extends ExternalIdentifiers {
+  ExternalIdentifiersLong() { this.getName().length() >= 32 }
+}
+
+predicate notSame(ExternalIdentifiers d, ExternalIdentifiers d2) {
+  not d = d2 and
+  d.getLocation().getStartLine() >= d2.getLocation().getStartLine()
+}
+
+from ExternalIdentifiers d, ExternalIdentifiers d2
+where
+  not isExcluded(d, Declarations1Package::externalIdentifiersNotDistinctQuery()) and
+  notSame(d, d2) and
+  if d instanceof ExternalIdentifiersLong and d2 instanceof ExternalIdentifiersLong
+  then d.getName().substring(0, 30) = d2.getName().substring(0, 30)
```

Comment:
Can you document why this is 31 characters (e.g. due to C99), and note that we don't currently support C90's requirements here?

Reply:
done!